### PR TITLE
Add numpy argpartition function support

### PR DESCRIPTION
**Pull request by rragundez**
Add numpy argpartition function support. 

I used some of the implementation of `partition`. Something to note it that I modified the already existing `_partition_factory` to add an optional argument if the user would need `np.argpartition` instead of `np.partition`, as you will see in the code it doesn't look the most pretty, but it was either that or write a `_argpartition_factory` function which has exactly the same logic it just swaps index elements at the same time it swaps the array with the values being sorted.

Let me know if you prefer to leave it as is or go the duplication route. I think this is the best compromise to not repeat code and have a single source of truth of partition logic.

Solves #2445
#4074 would need to be updated.

cheers

### CODE REVIEW

**Review by kc611 — APPROVED** [review]
Hi @rragundez , Thank you for the PR :smile: . Your contributions to Numba are highly appreciated. 

I gave the PR an initial review and it mostly LGTM!

A few points to note:
1. The logic does seem a bit off within the `_partition_factory`/`_partition`, but as you mentioned and from a maintainer's point of view, it's less code to manage and reuses existing logic so it does makes sense to keep it this way.

2. The `np.argpartition`, based on the existing `np.partition` logic does seem to deviate from NumPy behaviour, they are implemented using different algorithms (see https://github.com/numba/numba/pull/3320#issuecomment-422747016 for original `np.partition` implementation discussion) and whilst this might seem a bit implicit, since the ordering of the elements in the two partitions is supposed to be undefined, but it might be a good idea to mention the behaviour in the docs that the outputs may deviate from NumPy.

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1477]:
```suggestion
    def _partition(A, low, high, I=None):
```

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1479]:
```suggestion
        argpartition = True if I is None else False
        if argpartition:
            assert len(I) == len(A) # The corresponding index array should be valid for the array provided.
```

Just to make sure every upstream function uses the factory correctly.

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1531]:
```suggestion
    def _select(arry, k, low, high, idx=None):
```

**Review by kc611 — CHANGES_REQUESTED** [review]
A small suggestion regarding default function arguments. 

On this note is there any particular reason you went with the default argument for `idx` being `np.arange(0)` ?

**Comment by rragundez** [comment]
@kc611  I also do not like it at all, but since the function is passed to `register_jitable` it needs to make sense of the types of the variables and if left with default as  `None` it cannot make sense of the `None` type and the array of type `int64` which `I` or `idx` can become. So I decided to use the `np.arange(0)` as default value which is an empty array of type `int64`. 
Does that make sense? Let me know what you think.

**Comment by kc611** [comment]
Hi @rragundez, apologies for the delay. Following a small internal discussion, it'd be better for us to go with `None` as default argument. The reason being having a mutable default arguments for Python functions is generally not recommended due to the kind of unexpected behaviours it leads to down the line.

Additionally, the type inference within Numba is expected to figure out if the argument type is `None` and prune the branches accordingly during compile time.

**Comment by rragundez** [comment]
@kc611 Yeap, agree with the Python anti-pattern. I will make the changes and let's see what the the CI says, hopefully I made come mistake when testing and it will be able to resolve the `None` type.

**Comment by rragundez** [comment]
@kc611  done, please review. Thanks.

**Comment by rragundez** [comment]
@kc611 By the way I just remembered that indeed is a bad practice to set default parameters that are mutable but even though numpy arrays are mutable, an array of size 0 is immutable because it has no elements to change, and numpy arrays are of fixed size, therefore the default value of an int array of size zero is immutable. Let me know if this makes sense and how you would like me to proceed.

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1476]:
```suggestion
def _partition_factory(pivotimpl, argpartition=False):
```

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1479]:
```suggestion
```

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1528]:
```suggestion
_argpartition_w_nan = register_jitable(_partition_factory(nan_aware_less_than, argpartition=True))
```

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1545]:
```suggestion

_arg_select_w_nan = register_jitable(_select_factory(_argpartition_w_nan))
```

File: numba/np/arraymath.py
Comment (kc611) [inline numba/np/arraymath.py:1839]:
```suggestion
            _arg_select_w_nan(arry, kth, low, high, idx_arry)
```

**Review by kc611 — CHANGES_REQUESTED** [review]
Hi @rragundez , Thank you for making the changes. :smiley: . I think the reason for failure is that Numba isn't able to infer `I` as `None` during compile time and ends up trying to lower the branches with `argpartition` boolean. A way around this would be to make `argpartition` a compile time constant instead of runtime, this would make sure branch pruner absolutely removes the branches with the given boolean so that they aren't lowered into IR. 

I agree that an array of size 0 is immutable because it has no elements to change, but we can actually change it's attributes making it a 'mutable as a structure'. :slightly_smiling_face:  However, the issue at hand is that this practice is generally not recommended and even more so in Numba because it might lead to weird and unexpected IR issues down the line, even if it seems to work just fine right now.

**Comment by rragundez** [comment]
@kc611 thanks for the feedback. I made the requested changes and the CI is passing now. Please review.

**Review by kc611 — APPROVED** [review]
Thanks @rragundez, these changes LGTM!

**Comment by rragundez** [comment]
@kc611  Thanks for the approval. I cannot merge though, still says "Merging is blocked Merging can be performed automatically with 1 approving review."

**Comment by rragundez** [comment]
Hi @kc611, can you help me understand if this PR will be able to be merged? Thanks.

**Pull request merged**
